aten/src/ATen/core/TensorImpl.h

```diff
@@ -220,5 +312,753 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
+    reserved_ = true;
+    sizes_ = newDims;
+    numel_ = newNumel;
+  }
+
+  /**
+   * @brief Shrinks the outer-most dimension to given size, keeping the data.
+   *
+   * This method guarantees that no re-allocations are carried out, which means
+   * that the extra capacity after the end of the shurnk tensor is maintained.
+   */
+  void ShrinkTo(int64_t outer_dim) {
+    CAFFE_ENFORCE_WITH_CALLER(
+        is_contiguous_,
+        "Right now ShrinkTo is only supported on contiguous Tensor.");
+    CAFFE_ENFORCE_WITH_CALLER(sizes_.size() >= 1, "Tensor must be at least 1D");
+    CAFFE_ENFORCE_WITH_CALLER(
+        outer_dim <= sizes_[0],
+        "New outer dimension must be smaller than current.");
+    CAFFE_ENFORCE(
+        storage_.unique(),
+        "Can't call ShrinkTo on shared storage, please call Resize instead.");
+    sizes_[0] = outer_dim;
+    numel_ = std::accumulate(
+        sizes_.begin(),
+        sizes_.end(),
+        static_cast<int64_t>(1),
+        std::multiplies<int64_t>());
+  }
+
+  /**
+   * @brief Reserve space for the underlying tensor.
+   *
+   * This must be called after Resize(), since we only specify the first
+   * dimension This does not copy over the old data to the newly allocated space
+   */
+  template <class T>
+  void ReserveSpace(const T& outer_dim) {
+    CAFFE_ENFORCE_WITH_CALLER(
+        is_contiguous_,
+        "Right now ReserveSpace is only supported for contiguous Tensor.");
+    CAFFE_ENFORCE(
+        numel_ != -1, "size should be initialized before calling ReserveSpace");
+    CAFFE_ENFORCE(
+        storage_.unique(), "Can't call ReserveSpace on shared storage.");
+    auto newCapacity = sizes_;
+    newCapacity[0] = outer_dim;
+    auto newNumel = std::accumulate(
+        newCapacity.begin(),
+        newCapacity.end(),
+        static_cast<int64_t>(1),
+        std::multiplies<int64_t>());
+    if (newNumel * storage_.itemsize() <= storage_.capacity()) {
+      return;
+    }
+    // Old data is discarded
+    storage_.data_ptr().clear();
+    auto oldSize = numel_;
+    auto oldDims = sizes_;
+    Resize(newCapacity);
+    // Allocate new memory but don't copy over the data
+    raw_mutable_data(data_type_);
+    sizes_ = oldDims;
+    numel_ = oldSize;
+    reserved_ = true;
+  }
+
+  /**
+   * @brief Resizes a tensor.
+   *
+   * Resize takes in a vector of ints specifying the dimensions of the tensor.
+   * You can pass in an empty vector to specify that it is a scalar (i.e.
+   * containing one single item).
+   *
+   * The underlying storage may be deleted after calling Resize: if the new
+   * shape leads to a different number of items in the tensor, the old memory
+   * is deleted and new memory will be allocated next time you call
+   * mutable_data(). However, if the shape is different but the total number of
+   * items is the same, the underlying storage is kept.
+   */
+  template <typename... Ts>
+  void Resize(Ts... dim_source) {
```

Comment:
do all of these actually need to be on TensorImpl?  I buy Resize/raw_mutable_data, etc. but many of these seem like they can just be implementations on Tensor.  Thoughts?

Reply:
Moving these out of TensorImpl seems plausible, though once again, you're just kicking the can down the road for when Tensor class merges. In some cases we'll need to add new accessors to TensorImpl because it doesn't support enough fiddly bits to implement what you need (setters and also some private state, like `reserved_`).

There's also an order of operations question. It seems better to move these around after TensorImpl merges.